apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Trace/SimpleManagedTracer.cs

```diff
@@ -158,16 +159,19 @@ public T RunInSpan<T>(Func<T> func, string name, StartSpanOptions options = null
         /// <inheritdoc />
         public void EndSpan()
         {
-            lock (_stackLock)
-            {
-                CheckStackNotEmpty();
+            var currentStack = TraceStack;
+            CheckStackNotEmpty(currentStack);
 
-                TraceSpan span = _traceStack.Pop();
-                span.EndTime = Timestamp.FromDateTime(DateTime.Now.ToUniversalTime());
+            TraceSpan span;
+            currentStack = currentStack.Pop(out span);
```

Comment:
I don't believe so. As it is an immutable stack, the Pop call above doesn't nothing to the receiver. Rather it returns a new stack without the top item. The line below does change the stack currently stored in the AsyncLocal/logical call context, but that is a per-thread/per-async block storage location, which means there is no chance of race condition.